samples/nrf9160/serial_lte_modem/src/mqtt/th_mqtt_service.c

```diff
@@ -0,0 +1,598 @@
+	}
+#else
+	client.user_name = NULL;
+#endif
+#ifdef CONFIG_THIN_MQTT_BROKER_PASSWORD
+	if (strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD) > 0) {
+		client.password = k_malloc(sizeof(struct mqtt_utf8));
+		__ASSERT_NO_MSG(client.password != 0);
+		client.password->utf8 = k_malloc(64);
+		__ASSERT_NO_MSG(client.password->utf8 != 0);
+		memcpy(client.password->utf8,
+			(u8_t *)CONFIG_THIN_MQTT_BROKER_PASSWORD,
+			strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD));
+		client.password->size =
+			strlen(CONFIG_THIN_MQTT_BROKER_PASSWORD);
+	} else {
+		client.password = NULL;
+	}
+#else
+	client.password = NULL;
+#endif
+	client.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	client.rx_buf = rx_buffer;
+	client.rx_buf_size = sizeof(rx_buffer);
+	client.tx_buf = tx_buffer;
+	client.tx_buf_size = sizeof(tx_buffer);
+
+	/* MQTT transport configuration */
+	client.transport.type = MQTT_TRANSPORT_NON_SECURE;
+	return 0;
+}
+
+static void uninit_mqtt_client(void)
+{
+	client.broker = NULL;
+	if (client.user_name) {
+		k_free(client.user_name->utf8);
+		k_free(client.user_name);
+	}
+	if (client.password) {
+		k_free(client.password->utf8);
+		k_free(client.password);
+	}
+	fds.fd = INVALID_FDS;
+	k_sem_reset(&mqtt_connected_sem);
+	mqtt_connected = false;
+}
+
+int do_mqtt_connect(const u8_t *client_id, u8_t length)
+{
+	int err;
+
+#if defined(CONFIG_THIN_MQTT_USE_IP)
+	LOG_INF("connect to %s:%d",
+		log_strdup(CONFIG_THIN_MQTT_BROKER_HOST_IP),
+		CONFIG_THIN_MQTT_BROKER_PORT);
+#else
+	LOG_INF("connect to %s:%d",
+		log_strdup(CONFIG_THIN_MQTT_BROKER_HOST_NAME),
+		CONFIG_THIN_MQTT_BROKER_PORT);
+#endif
+
+	err = init_mqtt_client(client_id, length);
+	if (err) {
+		return err;
+	}
+
+	err = mqtt_connect(&client);
+	if (err) {
+		LOG_ERR("ERROR: mqtt_connect %d", err);
+		return err;
+	}
+
+	err = fds_init(&client);
+	if (err) {
+		LOG_ERR("ERROR: fds_init %d", err);
+		return err;
+	}
+
+	/** start polling now for CONNACK */
+	if (err == 0) {
+		k_sem_give(&mqtt_connected_sem);
+		mqtt_connected = true;
+	}
+
+	return err;
+}
+
+int do_mqtt_disconnect(void)
+{
+	int err = mqtt_disconnect(&client);
+
+	if (err) {
+		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+	}
+	uninit_mqtt_client();
+	return err;
+}
+
+/**@brief Function to subscribe to the configured topic
+ */
+int do_mqtt_subscribe(u8_t qos, const u8_t *topic, u8_t length)
+{
+	struct mqtt_topic subscribe_topic = {
+		.topic = {
+			.utf8 = (u8_t *)topic,
+			.size = length
+		},
+		.qos = qos
+	};
+
+	struct mqtt_subscription_list subscription_list = {
+		.list = &subscribe_topic,
+		.list_count = 1
+	};
+
+	subscription_list.message_id = sys_rand32_get();
+	LOG_INF("SUB with QoS %d (id %u)",
+		subscribe_topic.qos,
+		subscription_list.message_id);
+	LOG_DBG(" topic: %s", log_strdup(topic));
+	return mqtt_subscribe(&client, &subscription_list);
+}
+
+/**@brief Function to unsubscribe to the configured topic
+ */
+int do_mqtt_unsubscribe(const u8_t *topic, u8_t length)
```

Comment:
Why do you declare `*topic` as `const` and then down cast it on line 465?
If it needs indeed to be constant, down casting is dangerous 💣 
If it doesn't need to be constant, you can remove the `const` and the cast on line 465.